cpp/src/arrow/util/decimal.cc

```diff
@@ -521,6 +726,23 @@ std::string Decimal128::ToString(int32_t scale) const {
   return str;
 }
 
+static inline void ShiftAndAdd(std::string_view input, uint32_t* out) {
```

Comment:
This looks like a copy-paste job but I don't understand why it's useful. Converting a string to decimal32 or decimal64 shouldn't require multiple passes, so this should definitely be simplified.

Reply:
simplified this for decimal32 (decimal64 is able to just use the existing implementation used for the other types by just using a size of 1)